File: SrdDatabase/ClientApp/src/store/archdeaconry/details.ts
import { Reducer } from 'redux';
import { Action, AppThunkAction } from '..';
import { get } from '../../helpers/apiHelpers';
import { ArchdeaconryDetails } from '../../models/archdeaconry';
import { pagedResultsDefaults } from '../../models/shared';

const REQUEST_DETAILS = 'ARCHDEACONRY.REQUEST_DETAILS';
const RECEIVE_DETAILS = 'ARCHDEACONRY.RECEIVE_DETAILS';

const requestDetailsAction = () => ({
    type: REQUEST_DETAILS,
});

const receiveDetailsAction = (details: ArchdeaconryDetails) => ({
    type: RECEIVE_DETAILS,
    value: details,
});

const loadDetails = (id: number): AppThunkAction<Action> => (dispatch) => {
    dispatch(requestDetailsAction());

    get<ArchdeaconryDetails>(`api/archdeaconry/details/${id}`)
        .then(details => {
            dispatch(receiveDetailsAction(details));
        });
}

export const actionCreators = {
    loadDetails,
};

export interface State {
    detailsLoading: boolean;
    details: ArchdeaconryDetails;
}

const initialState: State = {
    detailsLoading: true,
    details: {
        archdeaconry: {},
        parishResults: { ...pagedResultsDefaults, parishes: [] },
        congregationResults: { ...pagedResultsDefaults, congregations: [] },
        eventResults: { ...pagedResultsDefaults, events: [] },
        paymentResults: { ...pagedResultsDefaults, payments: [] },
        chargeResults: { ...pagedResultsDefaults, charges: [] },
        censusResults: { ...pagedResultsDefaults, censuses: [] },
    },
};

export const reducer: Reducer<State, Action> = (state: State = initialState, action: Action): State => {
    switch (action.type) {
        case REQUEST_DETAILS:
            return {
                ...state,
                detailsLoading: true,
            };
        case RECEIVE_DETAILS:
            return {
                ...state,
                detailsLoading: false,
                details: action.value,
            };
        default:
            return state;
    }
};

File: SrdDatabase/Models/Archdeaconries/Details.cs
using System.Collections.Generic;
using SrdDatabase.Models.Congregations;
using SrdDatabase.Models.Events;
using SrdDatabase.Models.Parishes;

namespace SrdDatabase.Models.Archdeaconries
{
    public class Details
    {
        public Archdeaconry Archdeaconry { get; }

        public IEnumerable<Parish> Parishes { get; }

        public IEnumerable<Congregation> Congregations { get; }

        public IEnumerable<Event> RecentEvents { get; }

        public Details(
            Archdeaconry archdeaconry,
            IEnumerable<Parish> parishes,
            IEnumerable<Congregation> congregations,
            IEnumerable<Event> recentEvents)
        {
            Archdeaconry = archdeaconry;
            Parishes = parishes;
            Congregations = congregations;
            RecentEvents = recentEvents;
        }
    }

}
